selectAllExcept() pull request. #11025 Select columns from a table exc…

T0mats0 opened the pull request:
This is a pull request for Add selectAllExcept method #11025.

I attempted to incorporate the suggestions and discussions made previously. 

Lorenzo: first arg is a Table.
Story: provide tests for multiple calls.

If you call this method more than once someone might expect them to stack but the way this PR is done they don't stack.

Commit:
selectAllExcept() pull request. #1125 Select columns from a table except columns in exclusion list.

inoas commented on src/ORM/Query.php:
```diff
@@ -194,6 +195,30 @@ public function select($fields = [], $overwrite = false)
+            $table = $table->getTarget();
+        }
+        $aliasedFields = [];
+        if ($table instanceof Table) {
```
Maybe if the target is not a Table throw an Exception instead?

dereuromark replied:
Does it have to be a concrete table? Don't we usually try to program against interfaces for SOLID?

ADmad replied:
Only `Table` class has `getSchema()` method so checking for `RepositoryInterface()` isn't feasible.

This also makes me wonder if it's right to have a method on `Query` with takes an `Association` or `Table` as argument.

inoas replied:
Maybe it should take a some Interface that both implement?

xaphalanx replied:
Lorenzo recommended that the first argument be Table so there would be no confusion about what list of columns the user was selecting. (See original thread https://github.com/cakephp/cakephp/pull/11025).  

For now, I've left the parameters alone.  We could type hint Table and make the caller know to caller know that they have to call getTarget() on their association. The good news is the doc blocks will tell me what to pass.

It now throws an exception if something other than a table or association is passed.
Updated documentation and tests as requested.

inoas commented on src/ORM/Query.php:
```diff
@@ -194,6 +195,30 @@ public function select($fields = [], $overwrite = false)
+            $aliasedFields = $this->aliasFields($fields);
+        }
+
+        return $this->select($aliasedFields, $overwrite);
```
I don't see tests that toggle around overwrite, do you think adding them might be helpful?

inoas reviewed (changes requested):
$table should really be an instance of Table (or Association, and then table via getTarget) anything else I'd expect an Exception.

dereuromark commented on src/ORM/Query.php:
```diff
@@ -194,6 +195,30 @@ public function select($fields = [], $overwrite = false)
+     * Excluded fields should not be aliased names.
+     *
+     *
+     * @param Table|Association $table The table to use to get an array of columns
```
Classes must be FQCN here in doc blocks.

InfosecGene committed:
Updated doc block with FCQN
Updated test to include a reset being passed
Add a test to catch invalid argument exception.

xaphalanx commented:
Lorenzo recommended that the first argument be Table so there would be no confusion about what list of columns the user was selecting. (See original thread https://github.com/cakephp/cakephp/pull/11025).  

For now, I've left the parameters alone.  We could type hint Table and make the caller know to caller know that they have to call getTarget() on their association. The good news is the doc blocks will tell me what to pass.

It now throws an exception if something other than a table or association is passed.
Updated documentation and tests as requested.

InfosecGene committed:
Clearing whitespace conflict

InfosecGene committed:
Clearing end of file conflict.

stickler-ci committed:
Fixing style errors.

InfosecGene committed:
Merged 3.next updates

InfosecGene committed:
Merge branch 'selectAllExcept-1125' of https://github.com/xaphalanx/cakephp into selectAllExcept-1125

markstory commented on tests/TestCase/ORM/QueryTest.php:
```diff
@@ -3548,4 +3562,142 @@ public function testNotMatchingNested()
+    public function testSelectAllExceptWithMulitpleCalls()
+    {
+        $table = $this->getTableLocator()->get('Articles');
+        $this->getTableLocator()->get('Articles');
```
This line looks unnecessary.

markstory commented on tests/TestCase/ORM/QueryTest.php:
```diff
@@ -3548,4 +3562,142 @@ public function testNotMatchingNested()
+    public function testSelectAllExceptWithContains()
+    {
+        $table = $this->getTableLocator()->get('Articles');
+        $this->getTableLocator()->get('Articles');
```
This line looks unnecessary.

markstory commented on tests/TestCase/ORM/QueryTest.php:
```diff
@@ -3548,4 +3562,142 @@ public function testNotMatchingNested()
+            'Articles__id' => 'Articles.id',
+            'Articles__author_id' => 'Articles.author_id',
+            'Articles__title' => 'Articles.title',
+            'Articles__published' => 'Articles.published',
```
Shouldn't published be excluded as it was in the `selectAllExcept` call on 3650?

xaphalanx replied:
These tests show what would happen on multiple calls. As far as I know things go into $this->_parts but don't come out unless you reset/clear it. So on the first call all the columns but the excluded list go into the query. On a second call you can't then remove something from the query parts without resetting. So these tests show that behavior. This test also shows the "reset" behavior. Does that answer your question?

markstory replied:
Ok. Could we document how multiple calls interact in the API doc block? Having that should help people understand the method better.

xaphalanx replied:
Will do. Thanks for the feedback.

markstory commented on tests/TestCase/ORM/QueryTest.php:
```diff
@@ -3548,4 +3562,142 @@ public function testNotMatchingNested()
+            $table
+                ->find()
+                ->selectAllExcept([], ['body']);
+        } catch (\InvalidArgumentException $e) {
```
We generally use `expectException(InvalidArgumentException::class)`

xaphalanx replied:
Thanks for the pro-tip.

Commit:
Updated tests with a few improvements

Commit:
Merge branch '3.next' into selectAllExcept-1125

Commit:
Removed test that was failing. Looks like it's in master but not 3.next.

Commit:
Updated documentation explaining the behavior when you call selectAllExcept 2x in the same query.
